File: Memory Game/src/com/snatik/matches/ui/TileView.java
package com.snatik.matches.ui;

import android.content.Context;
import android.graphics.Bitmap;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import com.snatik.matches.R;

public class TileView extends FrameLayout {

	private RelativeLayout mTopImage;
	private ImageView mTileImage;
	private boolean mFlippedDown = true;

	public TileView(Context context) {
		this(context, null);
	}

	public TileView(Context context, AttributeSet attrs) {
		super(context, attrs);
	}

	public static TileView fromXml(Context context, ViewGroup parent) {
		return (TileView) LayoutInflater.from(context).inflate(R.layout.tile_view, parent, false);
	}

	@Override
	protected void onFinishInflate() {
		super.onFinishInflate();
		mTopImage = (RelativeLayout) findViewById(R.id.image_top);
		mTileImage = (ImageView) findViewById(R.id.image);
	}

	public void setTileImage(Bitmap bitmap) {
		mTileImage.setImageBitmap(bitmap);
	}

	public void flipUp() {
		mFlippedDown = false;
		mTopImage.setVisibility(View.GONE);
	}

	public void flipDown() {
		mFlippedDown = true;
		mTopImage.setVisibility(View.VISIBLE);
	}
	
	public boolean isFlippedDown() {
		return mFlippedDown;
	}
}

File: app/src/main/kotlin/com/snatik/matches/ui/game/TileView.kt
package com.snatik.matches.ui.game

import android.content.Context
import android.graphics.Bitmap
import android.view.LayoutInflater
import android.widget.FrameLayout
import androidx.core.view.isVisible
import com.snatik.matches.databinding.TileViewBinding

/** One card: a face-down back and the picture underneath, switched with a 3D flip. */
class TileView(context: Context) : FrameLayout(context) {

    private val binding = TileViewBinding.inflate(LayoutInflater.from(context), this)

    var isFaceUp: Boolean = false
        private set

    fun setImage(bitmap: Bitmap) = binding.image.setImageBitmap(bitmap)

    fun flipUp() {
        if (isFaceUp) return
        isFaceUp = true
        startAnimation(FlipAnimation(from = binding.imageTop, to = binding.image, forward = true))
    }

    fun flipDown() {
        if (!isFaceUp) return
        isFaceUp = false
        startAnimation(FlipAnimation(from = binding.image, to = binding.imageTop, forward = false))
    }

    /** Shows the picture immediately, used when the board is rebuilt mid-round. */
    fun showFaceUp() {
        isFaceUp = true
        binding.imageTop.isVisible = false
        binding.image.isVisible = true
    }
}
